### PR TITLE
nd-rolling

### LINKED ISSUE #4196: Convolution operation
I would like to perform 2d convolution operation with dask-data stored in xarray, defining the kernel.

Related issue are:
https://github.com/pydata/xarray/issues/1142
https://github.com/pydata/xarray/issues/2010
https://github.com/pydata/xarray/issues/1323

Referring to Dask, they have a specific image-oriented API which allows for convolutions:
http://image.dask.org/en/latest/dask_image.ndfilters.html

And this guy already implemented convolution operation based on xarray and dask:
https://github.com/serazing
https://github.com/serazing/xscale
https://xscale.readthedocs.io/en/latest/api.html

Are there plans to implement this feature?
In the meanwhile I'll try dask-image and xscale.

### PR DESCRIPTION
**Pull request by fujiisoup**
- [x] Closes #4196
 - [x] Tests added
 - [x] Passes `isort -rc . && black . && mypy . && flake8`
 - [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
 - [ ] New functions/methods are listed in `api.rst`

I noticed that the implementation of nd-rolling is straightforward.
The core part is implemented but I am wondering what the best API is, with keeping it backward-compatible.

Obviously, it is basically should look like
```python
da.rolling(x=3, y=3).mean()
```

A problem is other parameters, `centers` and `min_periods`.
In principle, they can depend on dimension. 
For example, we can have `center=True` only for `x` but not for `y`.

So, maybe we allow dictionary for them?
```python
da.rolling(x=3, y=3, center={'x': True, 'y': False}, min_periods={'x': 1, 'y': None}).mean()
```

The same thing happens for `.construct` method.
```python
da.rolling(x=3, y=3).construct(x='x_window', y='y_window', stride={'x': 2, 'y': 1})
```
I'm afraid if this dictionary argument was a bit too redundant.

Does anyone have another idea?

### CODE REVIEW

File: xarray/tests/test_dataarray.py [hunk 1]
```diff
@@ -6399,6 +6397,19 @@ def test_rolling_count_correct():
         assert_equal(result, expected)
 
 
+@pytest.mark.parametrize("da", (1,), indirect=True)
+@pytest.mark.parametrize("center", (True, False))
+@pytest.mark.parametrize("min_periods", (None, 1))
+def test_ndrolling_reduce(da, center, min_periods):
+    rolling_obj = da.rolling(time=3, a=2, center=center, min_periods=min_periods)
```
Comment (fujiisoup) [inline xarray/tests/test_dataarray.py:6404]: This is a basic test for the new nd-rolling

**Comment by max-sixty** [comment]
This looks very promising; I'm surprised it was possible without more code. I'm being slow, but where is the nd-rolling algo? I had thought bottleneck didn't support more than one dimension? https://bottleneck.readthedocs.io/en/latest/bottleneck.move.html, and that we'd have to implement our own in numbagg (which would be very possible)

**Comment by max-sixty** [comment]
Re the API, I think the dict is probably the best option, although it does complicate as the arguments become differently typed depending on one vs multiple dimensions.

One alternative is to allow fluent args, like:
```
(
    da
    .rolling(x=3, center=True)
    .rolling(y=5, min_periods=2)
    .mean()
)
```
...but does that then seem like the second `rolling` is operating on the _result_ of the first?

**Comment by fujiisoup** [comment]
Hi @max-sixty 

> One alternative is to allow fluent args, like:
> ...but does that then seem like the second rolling is operating on the result of the first?

I couldn't think of it until just now. But yes, it sounds to me like a repeated rolling operation.

> I'm being slow, but where is the nd-rolling algo? I had thought bottleneck didn't support more than one dimension? 

No. With nd-rolling, we need to use numpy reductions.
Its `skipna=True` operation is currently slow, but it can be improved replacing nan *before* the stride-trick.

**Comment by fujiisoup** [comment]
Another API concern. We now use `min_periods`, in which we implicitly assume one-dimension cases.

With nd-dimension, I think `min_counts` argument is more appropriate like bottleneck, which will limit the lower bound of the number of missing entries in the n-dimensional window.

Even if we leave it, we may disallow nd-argument of `min_periods` but keep it a scalar.

File: xarray/core/dask_array_ops.py [hunk 2]
```diff
@@ -32,6 +32,14 @@ def rolling_window(a, axis, window, center, fill_value):
     """
     import dask.array as da
 
+    # for nd-rolling.
+    # TODO It can be more efficient. Currently, the chunks at the boundaries
+    # will be copied, but it might be OK for many-chunked-arrays.
+    if hasattr(axis, "__len__"):
+        for ax, win, cen in zip(axis, window, center):
+            a = rolling_window(a, ax, win, cen, fill_value)
+        return a
```
Comment (fujiisoup) [inline xarray/core/dask_array_ops.py:41]: I hope it is true, though I didn't check it.
A suspicious part is `da.concatenate`, which I expect does not copy the original (strided-)array.

**Comment by dcherian** [comment]
I think the dictionary is OK. We can allow scalars for `center`, `stride` etc. in which case the scalar applies to all dimensions. This would preserve backward compatibility.

scalar for `min_counts` also seems OK to me. We can update it if someone requests it.

**Comment by fujiisoup** [comment]
I think now it is ready for review, though I'm sure tests miss a lot of edge cases.
Maybe we can fix them if pointed out.

**Comment by fujiisoup** [comment]
A possible improvement will be nan-reduction methods for nd-rolling.
Currently, we just use numpy nan-reductions, which is memory consuming for strided arrays.

This issue can be solved by replacing nan by appropriate values and applying nonnan-reduction methods, 
e.g.,
```python
da.rolling(x=2, y=3).construct(x='xw', y='yw').sum(['xw', 'yw'])
```
should be the same with
```python
da.rolling(x=2, y=3).construct(x='xw', y='yw', fill_value=0).sum(['xw', 'yw'], skipna=False)
```
and the latter is much more memory efficient.

I'd like to leave this improvement to future PR.

**Comment by fujiisoup** [comment]
I got an error for typechecking, only in CI but not in local, from the code that I didn't change.

**Comment by fujiisoup** [comment]
Could anyone kindly review this?

File: xarray/tests/test_dataset.py [hunk 3]
```diff
@@ -6007,6 +6005,48 @@ def test_rolling_reduce(ds, center, min_periods, window, name):
         assert src_var.dims == actual[key].dims
 
 
+@pytest.mark.parametrize("ds", (1,), indirect=True)
+@pytest.mark.parametrize("center", (True, False))
+@pytest.mark.parametrize("min_periods", (None, 1))
+@pytest.mark.parametrize("name", ("sum", "mean", "max"))
```
Comment (max-sixty) [inline xarray/tests/test_dataset.py:6011]: Will this test will fail for something like `std`? 
IIUC iteratively applying the function to the array for each dim only works for some functions.

Which is probably fine, no need to write more tests imo, though maybe worth making a note

**Comment by fujiisoup** [comment]
Thanks @max-sixty for the review ;)
I'll work for the update in a few days.

File: xarray/core/dask_array_ops.py (hunk 2 above)
Comment (fujiisoup) [reply xarray/core/dask_array_ops.py:41]: I noticed that ghosting breaks my expectation. I need to update the algo here.

File: xarray/tests/test_dataset.py (hunk 3 above)
Comment (fujiisoup) [reply xarray/tests/test_dataset.py]: In this case, many other functions should have the same result.
If there are some nans in random positions, the result should be different.
I added a few more tests and include function `std`, but I don't think it is necessary to test all nanops.

File: xarray/core/rolling.py [hunk 4]
```diff
@@ -251,13 +272,27 @@ def construct(self, window_dim, stride=1, fill_value=dtypes.NA):
 
         from .dataarray import DataArray
 
+        if window_dim is None:
+            if len(window_dim_kwargs) == 0:
+                raise ValueError(
+                    "Either window_dim or window_dim_kwargs need to be specified."
```
[the remark this answers is not in the document]
Comment (fujiisoup) [reply xarray/core/rolling.py:275]: For the backward-compatibility, we need to support a `str` for window_dim for 1d-rolling.
`either_dict_or_kwargs` does not support this situation.

File: xarray/tests/test_dataset.py [hunk 5]
```diff
@@ -6008,15 +6008,25 @@ def test_rolling_reduce(ds, center, min_periods, window, name):
...
-            ds.rolling(time=3, center=center, min_periods=min_periods), name
-        )().rolling(x=2, center=center, min_periods=min_periods),
+            ds.rolling(time=4, center=center, min_periods=min_periods), name
+        )().rolling(x=3, center=center, min_periods=min_periods),
+        name,
+    )()
+    assert_allclose(actual, expected)
+    assert actual.dims == expected.dims
+
+    # Do it in the oposite order
```
Comment (max-sixty) [inline xarray/tests/test_dataset.py:6025]:
```suggestion
    # Do it in the opposite order
```

**Comment by max-sixty** [comment]
I'm still a bit confused. How does the "roll over each dimension in turn" approach equal the "roll over both dimension together" approach with a function like std? Here's a proposed counter example:

```python
import xarray as xr
import numpy as np

da = xr.DataArray(np.asarray([[0,10,0],[0,10,0], [0,10,0]]), dims=list('xy'))
print(da)

<xarray.DataArray (x: 3, y: 3)>
array([[ 0, 10,  0],
       [ 0, 10,  0],
       [ 0, 10,  0]])
Dimensions without coordinates: x, y

x_std = da.rolling(dict(x=2)).std()
print(x_std)

<xarray.DataArray (x: 3, y: 3)>
array([[nan, nan, nan],
       [ 0.,  0.,  0.],
       [ 0.,  0.,  0.]])
Dimensions without coordinates: x, y

x_then_y_std = x_std.rolling(dict(y=2)).std()
print(x_then_y_std)

<xarray.DataArray (x: 3, y: 3)>
array([[nan, nan, nan],
       [nan,  0.,  0.],
       [nan,  0.,  0.]])
Dimensions without coordinates: x, y

combined_std = da.rolling(dict(x=2, y=2)).std()
print(combined_std)

<xarray.DataArray (x: 3, y: 3)>
array([[nan, nan, nan],
       [nan,  5.,  5.],
       [nan,  5.,  5.]])
Dimensions without coordinates: x, y

```

**Comment by dcherian** [comment]
I think it's more like
```
da.rolling(x=2).construct("xr").rolling(y=2).construct("yr").std(["xr", "yr"])
```

but that gives
```
array([[0., 5., 5.],
       [0., 5., 5.],
       [0., 5., 5.]])
```

**Comment by fujiisoup** [comment]
Thanks @max-sixty .

You are completely correct.
As the test pass, I was fooling myself. 

The reason was that the dataset I was using for the test does not have `time` and `x` simultaneously.
So I was not testing the 2d-rolling but just 1d-rolling.

Fixed. Now it correctly fails for `mean` and `std`, but passes with `max` and `sum`

**Comment by max-sixty** [comment]
Great! Thanks @fujiisoup ! 

Ready to go from my POV

**Comment by fujiisoup** [comment]
@max-sixty 
thanks for the review.
merged
